ring/util_test.go

```diff
@@ -232,15 +240,123 @@ func TestWaitRingStabilityShouldReturnErrorIfMaxWaitingIsReached(t *testing.T) {
+	t.Parallel()
+
+	const (
+		minStability = 2 * time.Second
+		maxWaiting   = 7 * time.Second
+	)
+
+	ring := createStartingRing()
+
+	done := addInstancesPeriodically(ring)
+	defer close(done)
+
+	startTime := time.Now()
+	require.Equal(t, context.DeadlineExceeded, WaitRingStability(context.Background(), ring, Reporting, minStability, maxWaiting))
+	elapsedTime := time.Since(startTime)
+
+	assert.GreaterOrEqual(t, elapsedTime.Milliseconds(), maxWaiting.Milliseconds())
+}
+
+func TestWaitRingTokensStability_ShouldReturnErrorIfInstancesAddedAndMaxWaitingIsReached(t *testing.T) {
+	t.Parallel()
+
+	const (
+		minStability = 2 * time.Second
+		maxWaiting   = 7 * time.Second
+	)
+
+	ring := createStartingRing()
+
+	done := addInstancesPeriodically(ring)
+	defer close(done)
+
+	startTime := time.Now()
+	require.Equal(t, context.DeadlineExceeded, WaitRingTokensStability(context.Background(), ring, Reporting, minStability, maxWaiting))
+	elapsedTime := time.Since(startTime)
+
+	assert.GreaterOrEqual(t, elapsedTime.Milliseconds(), maxWaiting.Milliseconds())
+}
+
+// Keep changing the ring in a way to avoid repeating the same set of states for at least 2 sec
+func changeStatePeriodically(ring *Ring) chan struct{} {
+	done := make(chan struct{})
+	go func() {
+		instanceToMutate := "instance-1"
+		states := []InstanceState{PENDING, JOINING, ACTIVE, LEAVING}
+		stateIdx := 0
+
+		for states[stateIdx] != ring.ringDesc.Ingesters[instanceToMutate].State {
+			stateIdx++
+		}
+
+		for {
+			select {
+			case <-done:
+				return
+			case <-time.After(time.Second):
+				stateIdx++
+				ring.mtx.Lock()
+				ringDesc := ring.ringDesc
+				desc := InstanceDesc{Addr: instanceToMutate, State: states[stateIdx%len(states)], Timestamp: time.Now().Unix()}
+				ringDesc.Ingesters[instanceToMutate] = desc
+				ring.mtx.Unlock()
+			}
+		}
+	}()
+
+	return done
+}
+
+func TestWaitRingStability_ShouldReturnErrorIfInstanceStateIsChangingAndMaxWaitingIsReached(t *testing.T) {
+	t.Parallel()
+
+	const (
+		minStability = 2 * time.Second
+		maxWaiting   = 7 * time.Second
+	)
+
+	ring := createStartingRing()
+
+	// Keep changing the ring.
+	done := changeStatePeriodically(ring)
+	defer close(done)
 
 	startTime := time.Now()
 	require.Equal(t, context.DeadlineExceeded, WaitRingStability(context.Background(), ring, Reporting, minStability, maxWaiting))
 	elapsedTime := time.Since(startTime)
 
-	assert.InDelta(t, maxWaiting, elapsedTime, float64(2*time.Second))
+	assert.GreaterOrEqual(t, elapsedTime.Milliseconds(), maxWaiting.Milliseconds())
+}
+
+func TestWaitRingTokensStability_ShouldReturnOnceMinStabilityOfInstancesHasBeenReachedWhileStateCanChange(t *testing.T) {
+	t.Parallel()
+
+	const (
+		minStability = 2 * time.Second
+		maxWaiting   = 7 * time.Second
+	)
+
+	ring := createStartingRing()
+
+	// Keep changing the ring.
+	done := changeStatePeriodically(ring)
+	defer close(done)
```

Comment:
Nit: This test leaves goroutine running even after test is finishes (for up to 1s). Other tests  have the same or similar problems (eg. leaving goroutines behind, if they fail early). Goleak took would flag such tests, even though goroutines finish soon after. Probably not worth fixing.

Reply:
huh? I'm confused, the done channel is there specifically to avoid that. How would it leak?

Reply:
Oh, somehow I read the code as if it was waiting 1s. My bad. The race between test end and goroutine exit is still there, but it's definitely not 1s.

Reply:
discussed offline, won't fix now